source/common/upstream/upstream_impl.cc

```diff
@@ -115,14 +116,25 @@ ClusterPtr ClusterImplBase::create(const Json::Object& cluster, ClusterManager&
 
   std::unique_ptr<ClusterImplBase> new_cluster;
   std::string string_type = cluster.getString("type");
+
+  auto selected_dns_resolver = dns_resolver;
+  if (cluster.hasObject("dns_resolvers")) {
+    auto resolver_addrs = cluster.getStringArray("dns_resolvers");
+    std::vector<Network::Address::InstanceConstSharedPtr> resolvers;
+    for (const auto& resolver_addr : resolver_addrs) {
+      resolvers.push_back(Network::Utility::parseInternetAddressAndPort(resolver_addr));
+    }
+    selected_dns_resolver = std::make_shared<Network::DnsResolverImpl>(dispatcher, resolvers);
```

Comment:
Can you add a comment to the effect of "We make this a shared pointer to deal with the distinct ownership scenarios that can exists with DNS resolvers. In one case, we have a single DNS resolver that is owned by the Server::Instance. In the other case, we have a per-cluster DNS resolver that is created here but ownership ultimately resides with StrictdnsClusterImpl/LogicalDnsCluster."